{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a36d8ac9",
   "metadata": {},
   "source": [
    "### Variant Analysis\n",
    "\n",
    "Project Overview:\n",
    "\n",
    "In this program, you have the ability to compare different virus/genomes to eachother in multiple different ways.\n",
    "\n",
    "Abilities:\n",
    "1. Genomic Stability:\n",
    "    Compute GC Content (%) and Molecular Weight.\n",
    "2. Protein Analysis:\n",
    "    Translate DNA to Protein and calculate Amino Acid frequencies.\n",
    "3. Visualization:\n",
    "    Compare variants using seaborn bar charts and heatmaps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f86dab6",
   "metadata": {},
   "source": [
    "## 1. Uploading FASTA Files\n",
    "\n",
    "Instructions:\n",
    "1. System will ask you to input varient/genome name (cosmetic).\n",
    "2. System will ask you to input file name (spelling and case sensitive)\n",
    "3. System will then ask if you would like to continue (Y/N only)\n",
    "\n",
    "If you enter the wrong filename, or specific file isnt in current working directory, it will give you and error and make you restart."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbb2b297",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "from Bio.SeqUtils import gc_fraction, molecular_weight\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Create a dictionary with varient names and file handles\n",
    "\n",
    "files = { }\n",
    "\n",
    "while True:\n",
    "    name=input('Enter the name of variant/genome:\\n>')\n",
    "    file=input('Enter the file name of variant/genome (Must be a fasta file):\\n>')\n",
    "    while str(file[-6:])!='.fasta':\n",
    "        file=input('Error! (Must be a fasta file)\\nPlease reenter:\\n>')\n",
    "    files[name]=file\n",
    "    iterate=input('Would you like to enter another? (Y/N)\\n>')\n",
    "    if iterate!='Y':\n",
    "        break\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e888d367",
   "metadata": {},
   "outputs": [],
   "source": [
    "data=[]\n",
    "\n",
    "# For loop to analyze each varient and create dataframe\n",
    "for variant, filename in files.items():\n",
    "    try:\n",
    "        record=SeqIO.read(filename,\"fasta\")\n",
    "    except FileNotFoundError:\n",
    "        print(f'Filename ({filename}) is incorrect, please restart and correct')\n",
    "\n",
    "    # Calculate GC Content (Genomic Stability) and Molecular Weight\n",
    "    gc_percentage = gc_fraction(record.seq) * 100\n",
    "    mol_weight = molecular_weight(record.seq, seq_type=\"DNA\")\n",
    "\n",
    "    # Get sequence length\n",
    "    length = len(record.seq)\n",
    "\n",
    "    # Translate DNA to Protein\n",
    "    proteinseq = record.seq.translate()\n",
    "\n",
    "    # Append to data list\n",
    "    data.append({\n",
    "        \"variant\":variant,\n",
    "        \"gc_percentage\":gc_percentage,\n",
    "        \"mol_weight\":mol_weight,\n",
    "        \"length\":length,\n",
    "        \"proteinseq\":str(proteinseq)\n",
    "    })\n",
    "\n",
    "# Create a Pandas dataframe from processed data\n",
    "df = pd.DataFrame(data)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1268b00a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create list to store dictionaries of counts for each varient\n",
    "aa_data=[]\n",
    "\n",
    "# Loop through each variant in the DataFrame\n",
    "for index,row in df.iterrows():\n",
    "    protein=row['proteinseq']\n",
    "    variant_name=row['variant']\n",
    "\n",
    "    # Initialize Dictionary with varient name\n",
    "    counts={}\n",
    "    counts[\"Variant\"] = variant_name\n",
    "\n",
    "    # Count frequency of each amino acid in sequence\n",
    "    for amino_acid in protein:\n",
    "        counts[amino_acid]=counts.get(amino_acid,0)+1 \n",
    "    aa_data.append(counts)\n",
    "\n",
    "# Create a Pandas Dataframe containing all amino acid counts per varient\n",
    "aa_df=pd.DataFrame(aa_data)\n",
    "aa_df=aa_df.set_index('Variant')\n",
    "aa_df=aa_df.fillna(0)\n",
    "\n",
    "# Drop stop codons to clean up dataframe\n",
    "aa_df=aa_df.drop(columns='*')\n",
    "\n",
    "aa_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9409a5ba",
   "metadata": {},
   "source": [
    "## 2. Comparative GC Content Visualization\n",
    "This visualization compares the GC content across variants. Through this we can understand that despite specific protein mutations, the overall stability of the genome is conserved."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "685f4268",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Seaborn barplot to compare GC content percentages across variants\n",
    "\n",
    "plt.figure(figsize=(6,6))\n",
    "sns.barplot(x='variant',y='gc_percentage',data=df)\n",
    "plt.title(\"GC Content Comparison\", fontsize=14)\n",
    "plt.ylabel(\"GC Content (%)\", fontsize=12)\n",
    "plt.xlabel(\"Variant\", fontsize=12)\n",
    "plt.ylim(37.5,38)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05be18c7",
   "metadata": {},
   "source": [
    "## 3. Comparison of Genomic Dimensions\n",
    "Here, I visualize the comparison of physical size of each genome using previously calculated molecular mass and nucleotide sequence length. Through this we observe that while all variants are structurally similar, there are slight variances in mass and length corresponding with specific mutational profiles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a98d117d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Seaborn barplot to compare sequence length across variants\n",
    "\n",
    "plt.figure(figsize=(6,6))\n",
    "sns.barplot(x='variant',y='length',data=df)\n",
    "plt.title(\"Sequence Length Comparison\", fontsize=14)\n",
    "plt.ylabel(\"Sequence Length\", fontsize=12)\n",
    "plt.xlabel(\"Variant\", fontsize=12)\n",
    "plt.ylim(29750,29800)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42cb88a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Seaborn barplot to sequence molecular weight across variants\n",
    "\n",
    "plt.figure(figsize=(6,6))\n",
    "sns.barplot(x='variant',y='mol_weight',data=df)\n",
    "plt.title(\"Molecular Weight Comparison\", fontsize=14)\n",
    "plt.ylabel(\"Molecular Weight\", fontsize=12)\n",
    "plt.xlabel(\"Variant\", fontsize=12)\n",
    "plt.ylim(9.20e6, 9.21e6)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc219cd1",
   "metadata": {},
   "source": [
    "## 4. Protein Analysis: Amino Acid Frequency\n",
    "By counting the frequency of each amino acid, we can observe shifts in the composition.\n",
    "\n",
    "Darker blue indicates a higher frequency of specific amino acids. Difference between rows highlight where mutations have altered the protein structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad40ba3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Seaborn heatmap to compare amino acid frequencies across variants\n",
    "\n",
    "plt.figure(figsize=(16,6))\n",
    "sns.heatmap(aa_df, cmap=\"Blues\", annot=True, fmt='g')\n",
    "plt.title(\"Amino Acid Frequency\", fontsize=16)\n",
    "plt.xlabel(\"Amino Acids\", fontsize=12)\n",
    "plt.ylabel(\"Variant\", fontsize=12)\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
